### PR TITLE
Support (EXPLAIN SELECT ...) as a subquery

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Support expression `(EXPLAIN SELECT ...)` in a subquery. Queries like `SELECT * FROM (EXPLAIN PIPELINE SELECT col FROM TABLE ORDER BY col)` became valid.

---

We have some tests that do `clickhouse client -q "EXPLAIN ..." | grep ...`, this PR would allow to write pure `.sql` tests for such cases.

https://github.com/ClickHouse/ClickHouse/blob/13de3d0665545ba6d34a8b1e65b7f7f6670f1eb5/tests/queries/0_stateless/02149_read_in_order_fixed_prefix_negative.sh#L15

https://github.com/ClickHouse/ClickHouse/blob/13de3d0665545ba6d34a8b1e65b7f7f6670f1eb5/tests/queries/0_stateless/01582_distinct_optimization.sh#L11

It allows to work with the result of EXPLAIN e.g in tests 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by kitaisreal** [comment]
I would prefer to add explain table function, because using EXPLAIN query as parameter breaks a lot of logic in parser (because EXPLAIN SELECT) is not an expression and it is impossible to analyze, because such EXPLAIN query need to be represented in query tree somehow.

**Comment by alexey-milovidov** [comment]
I would prefer to allow all query types that return results: EXPLAIN, DESCRIBE, SHOW, etc.

**Comment by kitaisreal** [comment]
I checked and in this pull request ```view(EXPLAIN ...)``` only change parser, so it is okay. Maybe it make sense to support such queries with rewritting on parser stage. It should not affect any other stages.
Example:
```
SELECT (EXPLAIN subquery); 
``` 
Result of parser level rewrite:
```
SELECT (SELECT explain_result FROM explainView(subquery)).
```
@alexey-milovidov, @vdimir what do you think ?

**Comment by kitaisreal** [comment]
Additional examples:
```
SELECT * FROM (DESCRIBE TABLE table)
```
```
SELECT * FROM (SELECT * FROM describeView(table))
```
```
SELECT * FROM (SHOW PROCESSLIST)
```
```
SELECT * FROM (SELECT * FROM processListView(table))
```
Maybe we can even create some generic view that can have argument that identifies its type 'PROCESS_LIST', 'DESCRIBE'.

File: src/TableFunctions/TableFunctionExplain.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        if (pulling_executor.pull(block))
+            blocks.push_back(std::move(block));
+        else
+            break;
+    }
+
+    if (blocks.size() == 1)
+        return blocks[0];
+
+    return concatenateBlocks(std::move(blocks));
```
Comment (vdimir) [inline src/TableFunctions/TableFunctionExplain.cpp:69]:
```suggestion
    return concatenateBlocks(blocks);
```

**Comment by vdimir** [comment]
@kitaisreal 

I've changed the implementation to support EXPLAIN subquery directly `SELECT (EXPLAIN subquery)`

It converts the query to `SELECT (SELECT * FROM explainView(subquery))` under the hood so that it would work with the analyzer without any changes. But calling directly `explainView` is impossible, and I suppose not expose it to the user. It requires more changes in the parser (like for the existing `view` function). Also, we should pass kind and settings into it, so it may start looking like `explainView('pipeline', query, header = 1, description = 1)`, which is difficult to parse and use, but it doesn't add any convenience.

If you are ok with implementation, let's finish this PR with `EXPLAIN` and I'll do the same for `SHOW`, `DESCRIBE`, and so on in the next.

**Comment by kitaisreal** [comment]
@vdimir great solution, I am okay with implementation. Let finish this for EXPLAIN and merge.

**Comment by vdimir** [comment]
#### CI failures on [`fa254bc`](https://github.com/ClickHouse/ClickHouse/pull/40630/commits/fa254bcec3934ea9b4db3fa65782bb31485538d1)

> Stateful tests (debug) — fail: 1, passed: 128, skipped: 2 [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stateful_tests__debug_.html)

https://github.com/ClickHouse/ClickHouse/issues/41238

> Stateless tests (aarch64) — fail: 1, passed: 4499, skipped: 19 [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stateless_tests__aarch64_.html)

`02267_file_globs_schema_inference`


> Stateless tests (debug, s3 storage) [1/3] — fail: 1, passed: 1500, skipped: 21 [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stateless_tests__debug__s3_storage__[1/3].html)

`00993_system_parts_race_condition_drop_zookeeper` - https://github.com/ClickHouse/ClickHouse/issues/36099


> Stateless tests (debug, s3 storage) [3/3] — Tests are not finished, fail: 1, passed: 1301, skipped: 10 [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stateless_tests__debug__s3_storage__[3/3].html)

`02397_system_parts_race_condition_drop_rm` https://github.com/ClickHouse/ClickHouse/issues/39022

> Stateless tests (release, s3 storage) — fail: 1, passed: 4484, skipped: 34 [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stateless_tests__release__s3_storage_.html)

`00993_system_parts_race_condition_drop_zookeeper`

> Stress test (asan) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt) [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stress_test__asan_.html)

```
2022.10.04 11:50:24.876635 [ 496080 ] {} <Error> test_3.rmt2 (ReplicatedMergeTreeAttachThread): Initialization failed, table will remain readonly. Error: The local set of parts of table test_3.rmt2 (8c49dc8d-da00-423e-9440-7907a80c88ee) doesn't look like the set of parts in ZooKeeper: 1.00 rows of 1.00 total rows in filesystem are suspicious. There are 1 uncovered unexpected parts with 1 rows (0 of them is not just-written with 0 rows), 0 missing parts (with 0 blocks), 0 covered unexpected parts (with 0 rows).
```

> Stress test (msan) — OOM killer (or signal 9) in clickhouse-server.log [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stress_test__msan_.html)

OOM

> Stress test (ubsan) — Killed by signal (in clickhouse-server.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/40630/fa254bcec3934ea9b4db3fa65782bb31485538d1/stress_test__ubsan_.html)

`Tried to lock part -3_6_6_0 for removal second time by ...` https://github.com/ClickHouse/ClickHouse/issues/39022

**Pull request merged**